### PR TITLE
refactor: router

### PR DESCRIPTION
**Pull request by notJoon**
# Description

## Refactor

- Refactored `route` file mainly.

## Remove

- Removed unused `compute_routes` and `gno_helper` files
- Removed underscore prefix from `_swap` functions and renamed them to `swapInner`
- Remove `DrySwapRoute` function and related methods

## Update

- Modified to use enum values instead of hardcoded `"EXACT_IN"`, `"EXACT_OUT"` strings. Additionally replace other hardcoded values with constants
- Add `_helper_test` file for testing purpose

Functions that use `Pool` currently doesn't have tests yet. This is because there's no way to directly access and initialize the pool's state for now.

### CODE REVIEW

File: router/router.gno [hunk 1]
```diff
@@ -139,76 +184,58 @@ func SwapRoute(
...
-			errInvalidInput.Error(), amountSpecified.ToString(),
-		)
-	}
-
-	if len(routes) < 1 || len(routes) > 7 {
-		return ufmt.Errorf(
-			"%s: route length must be 1~7. got %d",
-			errInvalidInput.Error(), len(routes),
-		)
+func DrySwapRoute(
```
Comment (dongwon8247) [inline router/router.gno:187]: Where is `DrySwapRoute` being used? Is this necessary?

Comment (notJoon) [reply router/router.gno:187]: Like `pool.DrySwap`, it seems to be used for simulation purposes.

Comment (dongwon8247) [reply router/router.gno:187]: @onlyhyde @r3v4s @notJoon can we check if these dry functions are being used in the transaction process? if so I want to understand how exactly they're used. I see `DrySwap`, `DrySwapRoute`, and `DryMint`, but add more if there are.

Comment (dongwon8247) [reply router/router.gno:232]: Delete `DrySwapRoute` if not being used.

**Review by dongwon8247 — COMMENTED** [review]
@notJoon Please add a unit test for `swap_multi.gno`, `swap_single.gno` and `wrap_unwrap.gno`.

> Functions that use Pool currently doesn't have tests yet. This is because there's no way to directly access and initialize the pool's state for now.

Can you create mockup data in `pool` that you can use for testing purposes in `router`? This seems essential tests.

File: router/router.gno (hunk 1 above)
Comment (notJoon) [reply router/router.gno:232]: removed https://github.com/gnoswap-labs/gnoswap/pull/431/commits/ae8bedebcd7b883617adbc5c4431062e8284eda9

File: router/utils.gno [hunk 2]
```diff
@@ -92,6 +94,20 @@ func min(a, b int) int {
...
+	return y
+}
+
+func i256Max(x, y *i256.Int) *i256.Int {
+	if x.Gt(y) {
+		return x
+	}
+	return y
+}
+
```
Comment (notJoon) [inline router/utils.gno:110]: needs to be moved to the `common` package. maybe `common/math`?

Comment (dongwon8247) [reply router/utils.gno:110]: How about we make `gnsmath `and `gnsutils `in the root folder and put math related files to `gnsmath`, util related files to `gnsutils`?

File: router/_helper_test.gno [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+	"gno.land/r/onbloc/bar"
+	"gno.land/r/onbloc/baz"
+	"gno.land/r/onbloc/foo"
+	"gno.land/r/onbloc/obl"
+	"gno.land/r/onbloc/qux"
+)
+
+const (
+	ugnotDenom string = "ugnot"
+	ugnotPath  string = "gno.land/r/gnoswap/v1/pool:ugnot"
```
Comment (r3v4s) [inline router/_helper_test.gno:24]:
```suggestion
	ugnotPath  string = "ugnot"
```

File: router/_helper_test.gno [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+
+// ----------------------------------------------------------------------------
+// ugnot
+
+func ugnotTransfer(t *testing.T, from, to std.Address, amount uint64) {
+	t.Helper()
+
+	std.TestSetRealm(std.NewUserRealm(from))
+	std.TestSetOrigSend(std.Coins{{ugnotDenom, int64(amount)}}, nil)
+	banker := std.GetBanker(std.BankerTypeRealmSend)
```
Comment (r3v4s) [inline router/_helper_test.gno:401]: gnolang/gno merged new pr relates to banker / coin issue
- https://github.com/gnolang/gno/pull/3239

need to check

File: router/protocol_fee_swap.gno [hunk 5]
```diff
@@ -11,14 +11,17 @@ import (
 	u256 "gno.land/p/gnoswap/uint256"
 )
 
+const (
+	defaultSwapFee = uint64(15) // 0.15%
```
Comment (r3v4s) [inline router/protocol_fee_swap.gno:15]: afaik, it should be value of bps
@onlyhyde need double check

File: router/protocol_fee_swap_test.gno [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+import (
+	"testing"
+
+	pusers "gno.land/p/demo/users"
+
+	u256 "gno.land/p/gnoswap/uint256"
+)
+
+func TestHandleSwapFee(t *testing.T) {
+	token0 := "token0"
```
Comment (r3v4s) [inline router/protocol_fee_swap_test.gno:12]: please use proper token path

File: router/utils_test.gno [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+			input:       "tokenA:tokenB",
+			shouldPanic: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			defer func() {
+				r := recover()
+				if (r != nil) != tt.shouldPanic {
```
Comment (r3v4s) [inline router/utils_test.gno:98]: use uassert.PanicsWithMessage to check panic with panic messsage

File: router/swap_inner_test.gno [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+    for _, tt := range tests {
+        t.Run(tt.name, func(t *testing.T) {
+            result := calculateSqrtPriceLimitForSwap(
+                tt.zeroForOne,
+                tt.fee,
+                tt.sqrtPriceLimitX96,
+            )
+            
+            if !result.Eq(tt.expected) {
```
Comment (r3v4s) [inline router/swap_inner_test.gno:55]: uassert.Equal

File: router/_helper_test.gno [hunk 9]
```diff
@@ -0,0 +1,450 @@
...
+
+	if ugnotBalanceOf(t, faucetAddress) < amount {
+		newCoins := std.Coins{{ugnotDenom, int64(amount)}}
+		ugnotMint(t, faucetAddress, newCoins[0].Denom, newCoins[0].Amount)
+		std.TestSetOrigSend(newCoins, nil)
+	}
+	ugnotTransfer(t, faucetAddress, to, amount)
+}
+
+func ugnotDeposit(t *testing.T, addr std.Address, amount uint64) {
```
Comment (r3v4s) [inline router/_helper_test.gno:443]: `func wrap` or `func ugnotWrap` seems more straightforward

File: router/utils.gno (hunk 2 above)
Comment (notJoon) [reply router/utils.gno:110]: related with https://github.com/gnoswap-labs/gnoswap/pull/444

**Comment by notJoon** [comment]
@onlyhyde The SwapRouter split commit was going into the staker (#437) instead of the router now that I checked, I'll revert this tomorrow.

File: router/protocol_fee_swap.gno (hunk 5 above)
Comment (onlyhyde) [reply router/protocol_fee_swap.gno]: yes i agree with you
we use bps unit for percent fee

Comment (notJoon) [reply router/protocol_fee_swap.gno]: https://github.com/gnoswap-labs/gnoswap/pull/431/commits/cb187e76b71b14df0a6703debb988a92ac1f437d

**Review by onlyhyde — APPROVED** [review]

File: router/protocol_fee_swap_test.gno (hunk 6 above)
Comment (r3v4s) [reply router/protocol_fee_swap_test.gno]: https://github.com/gnoswap-labs/gnoswap/blob/3f08c85c43ee0d9f89227801ceb7676b2b6d5575/router/protocol_fee_swap_test.gno#L50

File: router/utils_test.gno (hunk 7 above)
Comment (notJoon) [reply router/utils_test.gno:38]: this function is unused. removed https://github.com/gnoswap-labs/gnoswap/pull/431/commits/80468af96073f7240198d2e874a193b14a890228

File: router/swap_inner_test.gno (hunk 8 above)
Comment (notJoon) [reply router/swap_inner_test.gno]: https://github.com/gnoswap-labs/gnoswap/pull/431/commits/2b585a740a290e4010274860fca790a0eb61b785

**Review by onlyhyde — APPROVED** [review]

**Review by r3v4s — APPROVED** [review]
well done

**Pull request merged**